### PR TITLE
Add AssertingPartyMetadataRepository

### PR DESCRIPTION
**Pull request by jzheaux**
This PR introduces `OpenSamlAssertingPartyMetadataRepository`, a class that makes what is otherwise hidden behind `RelyingPartyRegistrations` configurable. For example, before this change, you would do:

```java
RelyingPartyRegistration.Builder builder = RelyingPartyRegistrations.fromMetadataLocation("uri");
```

With this change, you can do:

```java
var repository = OpenSamlAssertingPartyMetadataRepository.fromTrustedMetadataLocation("uri").build();
```

or

```java
var repository = new OpenSamlAssertingPartyMetadataRepository(myMetadataResolver);
```

or

```java
var repository = OpenSamlAssertingPartyMetadataRepository.withMetadataLocation("uri")
    .verificationCredentials((c) -> c.addAll(my, set, of, credentials)).build();
```

followed by:

```java
Collection<RelyingPartyRegistration.Builder> builders = StreamSupport.stream(repository.spliterator(), false)
    .map(RelyingPartyRegistration::withAssertingPartyMetadata);
```

`OpenSamlAssertingPartyMetadataRepository` uses an underlying `MetadataResolver` to refresh the metadata in an expiry-aware fashion. This can be used to back a `RelyingPartyRegistrationRepository` like so:

```java
@Component
public class RefreshableRelyingPartyRegistrationRepository implements IterableRelyingPartyRegistrationRepository {
    private final AssertingPartyMetadataRepository parties;

    // ...

    @Override 
    public RelyingPartyRegistration findByRegistrationId(String registrationId) {
        return applyRelyingParty(this.parties.findByEntityId(registrationId));
    }

    @Override 
    public Iterator<RelyingPartyRegistration> iterator() {
        return StreamSupport.stream(this.parties.spliterator(), false).map(this::applyRelyingParty).iterator();
    }

    private RelyingPartyRegistration applyRelyingParty(AssertingPartyMetadata metadata) {
        return RelyingPartyRegistration.withAssertingPartyMetadata(metadata)
            // apply relying party
            .build();
    }
}
```

### CODE REVIEW

**Comment by jzheaux** [comment]
@OrangeDog will you please take a look at this PR and see if it addresses the use cases you described in https://github.com/spring-projects/spring-security/issues/15090, https://github.com/spring-projects/spring-security/issues/15017, and https://github.com/spring-projects/spring-security/issues/15018 (related to https://github.com/spring-projects/spring-security/issues/12116)?

**Comment by OrangeDog** [comment]
- #15090 ❌ - I still need correct expiry-aware metadata refreshing, so will continue to adapt an `org.opensaml.saml.metadata.resolver.MetadataResolver` . Thus I still need a way to turn a single `EntityDescriptor` into a `RelyingPartyRegistration`.
- #15017 ✔- This looks like it works. I will have a look later to test it. However, what happens if the default encoding turns two different entity ids into the same registration id?
- #12116 ❓ - The default configuration needs to be secure (and the current lack of security should have a CVE filed). If untrusted metadata is provided (no signature from a remote source, no trust material available) it needs to fail. It should have the same behaviour as the deprecated implementation which spent a lot of time ensuring correct security. But for me, as I still need `MetadataResolver`, I will be getting that to do it for me (Shibboleth have spent even more time getting the security correct) rather than use your code.

File: docs/modules/ROOT/pages/servlet/saml2/login/overview.adoc [hunk 1]
```diff
@@ -639,6 +639,60 @@ class MyCustomSecurityConfiguration  {
...
+[source,java,role="primary"]
+----
+@Configuration
+@EnableWebSecurity
+public class MyCustomSecurityConfiguration {
+	@Bean
+    RelyingPartyRegistrationsDecoder registrationsDecoder() {
+		PublicKey pub = lookupAssertingPartyPublicKey();
+		Set<Credential> credentials = Set.of(new BasicCredential(pub));
+		return new OpenSamlRelyingPartyRegistrationsDecoder(credentials);
```
Comment (OrangeDog) [inline docs/modules/ROOT/pages/servlet/saml2/login/overview.adoc:659]: This constructor does not appear to exist.

**Comment by OrangeDog** [comment]
I've tested what I'm using (see above) and it's definitely an improvement 👍 . Please make `OpenSamlRelyingPartyRegistrationsDecoder.convert(EntityDescriptor)` public though.

For the signing, perhaps it should always use the filter, but with an empty/system trust store if none is provided. Then when decoding, set `requireSignedRoot` based on the URI scheme, or provide some customisation for what is trusted (with a safe default).

----

> With this change, applications can specify query parameters in the `authenticationRequestUri` value through a new method: `authenticationRequestUriQuery`:

Rather than parse a query string template, perhaps a method that just takes the parameter name would be better?

```
http
    .saml2Login((saml2) -> saml2
        .authenticationRequestUri("/saml/login")
        .authenticationRequestUriParam("peerEntityId")
    )
```

or 

```
http
    .saml2Login((saml2) -> saml2
        .authenticationRequestUri((uri) -> uri
            .path("/saml/login")
            .queryParam("peerEntityId")
        )
    )
```

**Comment by jzheaux** [comment]
> But for me, as I still need MetadataResolver, I will be getting that to do it for me (Shibboleth have spent even more time getting the security correct) rather than use your code.
> ...
>  Please make OpenSamlRelyingPartyRegistrationsDecoder.convert(EntityDescriptor) public though.

I think what may be better than this is enhancing `OpenSamlAssertingPartyDetails#withEntityDescriptor` to do the work of preparing the asserting party details instead of the decoder doing that. Then I think you'd not use the decoder at all.

**Comment by OrangeDog** [comment]
> I think what may be better than this is enhancing `OpenSamlAssertingPartyDetails#withEntityDescriptor` to do the work of preparing the asserting party details instead of the decoder doing that. Then I think you'd not use the decoder at all.

I don't care where that code is, as long as I can get at it without reflection hacks during my `EntityDescriptor` -> `RelyingPartyRegistration` conversion in my `findUniqueByAssertingPartyEntityId` adaptor.

**Comment by jzheaux** [comment]
Thanks for all the feedback and insight, @OrangeDog. I'm enjoying this collaboration with you.

Specifically, your comments about `MetadataResolver` inspired a different design that is hopefully more useful in the way of creating a refreshing, expiry-aware repository.

> Rather than parse a query string template, perhaps a method that just takes the parameter name would be better?

I'd like to keep the placeholder there as I think a nice feature one day would be to allow for other placeholders that are already supported in other parts of Spring Security like `{relyingPartyEntityId}` and `{assertingPartyEntityId}`. Given that, I think there is less cognitive load for the user to supply the query as a string, though I'd be open to doing something like:

```java
.authenticationRequestUri("/saml2/authenticate", "registrationId={registrationId}")
```

Such just didn't seem to buy very much since I'm already parsing the `=`.

> The default configuration needs to be secure (and the current lack of security should have a CVE filed). If untrusted metadata is provided (no signature from a remote source, no trust material available) it needs to fail.

I'd prefer to leave the defaults as they are, which means that if no credentials are provided, then no signature verification is performed. This is also what `MetadataResolver` in OpenSAML does by default.

I disagree that this is a CVE on the grounds that HTTPS is the standard for secure endpoints across Spring Security. When deviating from that standard, Security tries to make the secure option the easy one. But going further than that, we can easily get in the way -- for example. it's not true that every HTTP endpoint requires trust material as `http://localhost/` does not.

**Comment by jzheaux** [comment]
Also, in the meantime, I've addressed #15090 in the way [we already discussed](https://github.com/spring-projects/spring-security/pull/15349#issuecomment-2214757347) and committed that to `main`.

**Comment by OrangeDog** [comment]
> a different design that is hopefully more useful

Very interesting. I'll have a deeper look at that (and probably edit this before you wake up).

> HTTPS is the standard for secure endpoints

Not in SAML. That's why everything has XML-level signatures and optional encryption.

> it's not true that every HTTP endpoint requires trust material

That may be true, but disabling security should be a conscious choice by the developer, not the default.

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/OpenSamlAssertingPartyMetadataRepository.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+			@Override
+			public AssertingPartyMetadata next() {
+				return OpenSamlAssertingPartyDetails.withEntityDescriptor(descriptors.next()).build();
+			}
+		};
+	}
+
+	@Override
+	public AssertingPartyMetadata findByEntityId(String entityId) {
```
Comment (OrangeDog) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/OpenSamlAssertingPartyMetadataRepository.java:99]: This should have `@Nullable` to match the interface definition.
Then it should return `null` instead of throwing `NoSuchElementException`.

I used `resolveSingle` instead of `resolve` to make that simpler.

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/RelyingPartyRegistration.java [hunk 3]
```diff
@@ -338,6 +338,25 @@ public static Builder withAssertingPartyDetails(AssertingPartyDetails assertingP
...
+	 * Presented as a convenience method when working with
+	 * {@link AssertingPartyMetadataRepository} return values. As such, only supports
+	 * {@link AssertingPartyMetadata} instances of type {@link AssertingPartyDetails}.
+	 * @param metadata the metadata used to initialize the
+	 * {@link RelyingPartyRegistration} {@link Builder}
+	 * @return {@link Builder} to create a {@link RelyingPartyRegistration} object
+	 * @since 6.4
+	 */
+	public static Builder withAssertingPartyMetadata(AssertingPartyMetadata metadata) {
+		Assert.isInstanceOf(AssertingPartyDetails.class, "metadata must be of type AssertingPartyDetails");
```
Comment (OrangeDog) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/RelyingPartyRegistration.java:356]: This is checking the class of the message string, which always fails.

**Review by OrangeDog — CHANGES_REQUESTED** [review]

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/OpenSamlAssertingPartyMetadataRepository.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+	public OpenSamlAssertingPartyMetadataRepository(MetadataResolver metadataResolver) {
+		this.metadataResolver = metadataResolver;
+	}
+
+	@Override
+	@NonNull
+	public Iterator<AssertingPartyMetadata> iterator() {
+		CriteriaSet all = new CriteriaSet(new EntityRoleCriterion(IDPSSODescriptor.DEFAULT_ELEMENT_NAME));
+		Iterator<EntityDescriptor> descriptors = resolve(all).iterator();
```
Comment (OrangeDog) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/OpenSamlAssertingPartyMetadataRepository.java:84]: Whether this works depends on how the resolver is set up. If there's no index for `EntityRoleCriterion` then it will return nothing. Further, some resolvers e.g. `ChainingMetadataResolver` will not return all possible matches.

It can work with `EvaluableEntityRoleEntityDescriptorCriterion`, but only if the resolver has `resolveViaPredicatesOnly` set to `true`. 

You also cannot rely on `IterableMetadataSource` because not all classes implement that. You can special-case some classes to iterate over their delegates, but again not all of those may be iterable.

Basically, there's no way to implement this without restricting the configuration of resolvers that can be used. Looking for `IterableMetadataSource` is probably the better way.

Comment (OrangeDog) [reply saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/OpenSamlAssertingPartyMetadataRepository.java:84]: e.g.
```java
Stream<EntityDescriptor> allEntities(MetadataResolver resolver) {
    if (resolver instanceof IterableMetadataSource iterable) {
        return StreamSupport.stream(iterable.spliterator(), false);
    } else if (resolver instanceof ChainingMetadataResolver chaining) {
        return chaining.getResolvers().stream().flatMap(this::allEntities);
    } else if (resolver instanceof CompositeMetadataResolver composite) {
        return composite.getResolvers().stream().flatMap(this::allEntities);
    } else {
        return Stream.empty();
    }
}
```

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/RelyingPartyRegistration.java (hunk 3 above)
Comment (jzheaux) [reply saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/RelyingPartyRegistration.java]: Because the AssertingPartyMetadataRepository returns an interface, I'd like to move `RelyingPartyRegistration` towards using it as well. Good feedback though, I've updated the PR so that it doesn't need to do the `instanceof` check.

**Comment by OrangeDog** [comment]
I think this needs more test coverage, in particular the parts that didn't work that I pointed out above.
